server/schedule/operator/builder.go

```diff
@@ -183,99 +235,190 @@ func (b *Builder) SetLightWeight() *Builder {
 	return b
 }
 
+// SetForceTargetLeader marks the step of transferring leader to target is forcible. It is used for grant leader.
+func (b *Builder) SetForceTargetLeader() *Builder {
+	b.forceTargetLeader = true
+	return b
+}
+
+func (b *Builder) setUseJointConsensus(useJointConsensus bool) *Builder {
+	b.useJointConsensus = useJointConsensus
+	return b
+}
+
 // Build creates the Operator.
 func (b *Builder) Build(kind OpKind) (*Operator, error) {
+	var brief string
+
 	if b.err != nil {
 		return nil, b.err
 	}
 
-	brief, err := b.prepareBuild()
-	if err != nil {
-		return nil, err
+	if brief, b.err = b.prepareBuild(); b.err != nil {
+		return nil, b.err
 	}
 
-	kind, err = b.buildSteps(kind)
-	if err != nil {
-		return nil, err
+	if b.useJointConsensus {
+		kind, b.err = b.buildStepsWithJointConsensus(kind)
+	} else {
+		kind, b.err = b.buildStepsWithoutJointConsensus(kind)
+	}
+	if b.err != nil {
+		return nil, b.err
 	}
 
 	return NewOperator(b.desc, brief, b.regionID, b.regionEpoch, kind, b.steps...), nil
 }
 
 // Initialize intermediate states.
 func (b *Builder) prepareBuild() (string, error) {
-	var voterCount int
-	for _, p := range b.targetPeers.m {
-		if !core.IsLearner(p) {
+	b.toAdd = newPeersMap()
+	b.toRemove = newPeersMap()
+	b.toPromote = newPeersMap()
+	b.toDemote = newPeersMap()
+
+	voterCount := 0
+	for _, peer := range b.targetPeers {
+		if !core.IsLearner(peer) {
 			voterCount++
 		}
 	}
 	if voterCount == 0 {
 		return "", errors.New("cannot create operator: target peers have no voter")
 	}
 
-	// Diff `originPeers` and `targetPeers` to initialize `toAdd`,
-	// `toPromote`, `toRemove`.
-	for _, o := range b.originPeers.m {
-		n := b.targetPeers.Get(o.GetStoreId())
-		// no peer in targets, or target is learner while old one is voter.
-		if n == nil || (core.IsLearner(n) && !core.IsLearner(o)) {
+	// Diff `originPeers` and `targetPeers` to initialize `toAdd`, `toRemove`, `toPromote`, `toDemote`.
+	// Note: Use `toDemote` only when `useJointConsensus` is true. Otherwise use `toAdd`, `toRemove` instead.
+	for _, o := range b.originPeers {
+		n := b.targetPeers[o.GetStoreId()]
+		if n == nil {
 			b.toRemove.Set(o)
 			continue
 		}
-		if core.IsLearner(o) && !core.IsLearner(n) {
-			b.toPromote.Set(n)
+
+		// If the peer id in the target is different from that in the origin,
+		// modify it to the peer id of the origin.
+		if o.GetId() != n.GetId() {
+			n = &metapb.Peer{
+				Id:      o.GetId(),
+				StoreId: o.GetStoreId(),
+				Role:    n.GetRole(),
+			}
+		}
+
+		if core.IsLearner(o) {
+			if !core.IsLearner(n) {
+				// learner -> voter
+				b.toPromote.Set(n)
+			}
+		} else {
+			if core.IsLearner(n) {
+				// voter -> learner
+				if b.useJointConsensus {
+					b.toDemote.Set(n)
+				} else {
+					b.toRemove.Set(o)
+					// Need to add `b.toAdd.Set(n)` in the later targetPeers loop
+				}
+			}
 		}
 	}
```

Comment:
What about adding a TODO here? We can simply the code here in future.

Reply:
After the refactoring, `useJointConsensus` and `toDemote` are used repeatedly in the old logic. I think this is very difficult.

Reply:
Look forward to